### PR TITLE
Checks all custom actions for cycles. Fixes wixtoolset/issues#6201

### PR DESCRIPTION
**Pull request by cpuwzd**
@robmen,

This pull request is from my master branch. It is mergeable. It should have wound up on my Issue#6201 branch, but it didn't. I'd like to abandon the pull request(s) I've made on that branch and delete my local branch. I don't want to delete my local branch until I know that it won't mess things up worse than they are.

I'm missing a concept somewhere. It's like driving into oncoming traffic. If the upstream repo is updated while I'm working on a fix, I get into trouble when I do a rebase based on the upstream repo. This seems perhaps  especially true if I've done multiple local commits before I get to rebasing. I'm not sure what sequence of transactions are required to update my origin repo's master and project branches. another thing: the default remote for git push seems to be upstream, rather than origin, contrary to the documentation I've read. I know that you might have overridden the default in a configuration file.

The merge conflicts I've been seeing are cases where I have modified the same file on more than one local commit. I assume that the onus is on me to figure out which changes to omit from the rebase in order to avoid these conflicts.

I'd appreciate any pointers that might help clear this up.

Ron Martin

### CODE REVIEW

**Comment by rseanhall** [comment]
Change the default remote for the push operation - https://stackoverflow.com/questions/18801147/changing-the-git-remote-push-to-default.

Something you could have done instead of creating a new pull request:
1. Switch to the new branch with your changes: `git checkout master`
1. Rename your PR branch: `git branch -m issue#6201 issue#6201-old`
1. Recreate your PR branch from the current branch: `git checkout -b issue#6201`
1. Overwrite the PR branch from this new branch: `git push origin issue#6201 -f`

The general idea is you always commit to a feature branch, and not master. When you need to put your local changes on top of the latest from upstream (which should only be necessary if you're touching files that other people touched), switch to master and pull from upstream. Then switch back to your feature branch and rebase on master.

We would need to know the commands you're running (or the UI tool you're using) to know what you're doing wrong.

File: src/WixToolset.Core.WindowsInstaller/Bind/SequenceActionsCommand.cs [hunk 1]
```diff
@@ -562,64 +565,105 @@ private IEnumerable<string> GetRequiredActionIds()
...
-                if (WindowsInstallerStandard.TryGetStandardAction(parentActionKey, out parentActionSymbol))
+                if (!firstReference.ContainsKey(actionSymbol))
                 {
-                    // Create a clone to avoid modifying the static copy of the object.
-                    // TODO: consider this: parentActionSymbol = parentActionSymbol.Clone();
+                    firstReference[actionSymbol] = initialActionSymbol;
+                }
+                else if (firstReference[actionSymbol] == initialActionSymbol)
+                {
+                    throw new WixException(ErrorMessages.ActionCircularDependency(actionSymbol.SourceLineNumbers,
```
Comment (rseanhall) [inline src/WixToolset.Core.WindowsInstaller/Bind/SequenceActionsCommand.cs:600]: @robmen Why are we throwing exceptions here?

**Review by robmen — CHANGES_REQUESTED** [review]
Some style nits, some optimizations.

**Comment by robmen** [comment]
@cpuwzd when you're first working with git and GitHub I really do suggest creating a couple test repos with simple text files and a fork. Then do edits to force merge conflicts, especially across forks. I found that was one of the best way to learn git and GitHub mechanics. Until you get a solid grasp on git, it can be very frustrating to work in an active repo with others for two reasons: 1. you may feel like you're spending more time in git than working on code and 2. you could frustrate the repo owners because you're ending up with lots of "noisy commits" that have to get sorted out of your PR (or the PR simply won't rebase).

Use small repos of your own creation where you can destroy or corrupt the history without care.

File: src/WixToolset.Core.WindowsInstaller/Bind/SequenceActionsCommand.cs [hunk 2]
```diff
@@ -562,64 +565,105 @@ private IEnumerable<string> GetRequiredActionIds()
...
 
-            var parentActionName = (after ? actionSymbol.After : actionSymbol.Before);
-            var parentActionKey = actionSymbol.SequenceTable.ToString() + "/" + parentActionName;
-
-            if (!requiredActionSymbols.TryGetValue(parentActionKey, out var parentActionSymbol))
+            do
             {
-                // If the missing parent action is a standard action (with a suggested sequence number), add it.
-                if (WindowsInstallerStandard.TryGetStandardAction(parentActionKey, out parentActionSymbol))
+                if (!firstReference.ContainsKey(actionSymbol))
```
[the remark this answers is not in the document]
Comment (cpuwzd) [reply src/WixToolset.Core.WindowsInstaller/Bind/SequenceActionsCommand.cs:594]: How's this?
`
                if (!firstReference.TryGetValue(currentActionSymbol, out var existingInitialActionSymbol))
                {
                    firstReference[currentActionSymbol] = actionSymbol;
                }
                else if (existingInitialActionSymbol == actionSymbol)
                {
                    throw new WixException(ErrorMessages.ActionCircularDependency(currentActionSymbol.SourceLineNumbers, currentActionSymbol.SequenceTable.ToString(), currentActionSymbol.Action, previousActionSymbol.Action));
                }
`

File: src/WixToolset.Core.WindowsInstaller/Bind/SequenceActionsCommand.cs [hunk 3]
```diff
@@ -562,64 +565,105 @@ private IEnumerable<string> GetRequiredActionIds()
...
+                {
+                    throw new WixException(ErrorMessages.ActionCircularDependency(actionSymbol.SourceLineNumbers,
+                        actionSymbol.SequenceTable.ToString(), actionSymbol.Action, previousActionSymbol.Action));
+                }
+
+                if (null == actionSymbol.Before && null == actionSymbol.After)
+                {
+                    if (firstTime)
+                    {
+                        throw new InvalidOperationException(
```
[the remark this answers is not in the document]
Comment (cpuwzd) [reply src/WixToolset.Core.WindowsInstaller/Bind/SequenceActionsCommand.cs:608]: All exceptions in SequenceActionsCommand are now single line and I'm applying the same strategy elsewhere.

File: src/WixToolset.Core.WindowsInstaller/Bind/SequenceActionsCommand.cs [hunk 4]
```diff
@@ -562,64 +565,105 @@ private IEnumerable<string> GetRequiredActionIds()
...
+                            "Found an action with no Sequence, Before, or After column set.");
```
[the remark this answers is not in the document]
Comment (cpuwzd) [reply src/WixToolset.Core.WindowsInstaller/Bind/SequenceActionsCommand.cs:609]: I took a shot at this. I'm not sure how much validity checking is required on the action symbol property values. There are two occurrences of this problem in SequnceActionsCommand at this point. The first is in Execute. The second is in SequenceActionSymbol. They both look like this:

`
throw new InvalidOperationException($"Found an action [{actionSymbol.Id.Id}] at [{actionSymbol.SourceLineNumbers}] with no Sequence, Before, or After column set.");
`
I also stumbled across this in Messages:
`
        public static Message OverridableActionCollision2(SourceLineNumber sourceLineNumbers)
        {
            return Message(sourceLineNumbers, Ids.OverridableActionCollision2, "The location of the action related to previous error.");        }
`
My hands are full at the moment.

File: src/WixToolset.Core.WindowsInstaller/Bind/SequenceActionsCommand.cs [hunk 5]
```diff
@@ -562,64 +565,105 @@ private IEnumerable<string> GetRequiredActionIds()
...
+        /// initial action symbol, and thus is also not part of a cycle. However, any action symbol
+        /// encountered that has already been encountered starting from the same initial action symbol
+        /// is an indication that the current action symbol is part of a cycle.
+        /// </para>
         /// <param name="actionSymbol">The action symbol to be sequenced.</param>
         /// <param name="requiredActionSymbols">Collection of actions which must be included.</param>
-        private void SequenceActionSymbol(WixActionSymbol actionSymbol, Dictionary<string, WixActionSymbol> requiredActionSymbols)
+        /// <param name="firstReference">The first encountered action symbol that led to each action symbol.</param>
+        private void SequenceActionSymbol(WixActionSymbol actionSymbol,
+            Dictionary<string, WixActionSymbol> requiredActionSymbols,
```
[the remark this answers is not in the document]
Comment (cpuwzd) [reply src/WixToolset.Core.WindowsInstaller/Bind/SequenceActionsCommand.cs:582]: Added to my mental model of your formatting rules.

File: src/test/WixToolsetTest.CoreIntegration/WixToolsetTest.CoreIntegration.csproj [hunk 6]
```diff
@@ -12,6 +12,10 @@
     <Content Include="TestData\**" CopyToOutputDirectory="PreserveNewest" />
   </ItemGroup>
 
+  <ItemGroup>
+    <None Remove="TestData\CustomAction\CustomActionCycleWithTail.wxs" />
+  </ItemGroup>
```
[the remark this answers is not in the document]
Comment (cpuwzd) [reply src/test/WixToolsetTest.CoreIntegration/WixToolsetTest.CoreIntegration.csproj:17]: This ItemGroup did not appear when I recreated my Issue#6201 branch.

File: src/WixToolset.Core.WindowsInstaller/Bind/SequenceActionsCommand.cs [hunk 7]
```diff
@@ -562,64 +565,105 @@ private IEnumerable<string> GetRequiredActionIds()
...
+                    initialParentActionSymbol = parentActionSymbol ?? throw new InvalidOperationException(
+                        "Found an action with no Sequence, Before, or After column set.");
                 }
-            }
-            else if (actionSymbol == parentActionSymbol || this.ContainsChildActionSymbol(actionSymbol, parentActionSymbol)) // cycle detected
-            {
-                throw new WixException(ErrorMessages.ActionCircularDependency(actionSymbol.SourceLineNumbers, actionSymbol.SequenceTable.ToString(), actionSymbol.Action, parentActionSymbol.Action));
-            }
+
+                previousActionSymbol = actionSymbol;
```
[the remark this answers is not in the document]
Comment (cpuwzd) [reply src/WixToolset.Core.WindowsInstaller/Bind/SequenceActionsCommand.cs:623]: It's complicated. I wound up going back to an approach I had tried, but rejected. From the version you have seen, it amounts to unrolling the first pass of the loop and attaching a streamlined version of the loop to the first pass, just before it gives its final blessing. It's easier to read and cleaner.

File: src/WixToolset.Core.WindowsInstaller/Bind/SequenceActionsCommand.cs [hunk 8]
```diff
@@ -562,64 +565,105 @@ private IEnumerable<string> GetRequiredActionIds()
...
+                        "Found an action with a non-existent {0} action: {1}.",
+                        (after ? "After" : "Before"), parentActionName));
+                }
             }
 
-            return result;
+            return parentActionSymbol;
         }
+    
+        private bool GetAfter(WixActionSymbol actionSymbol) => (actionSymbol.After != null);
```
[the remark this answers is not in the document]
Comment (cpuwzd) [reply src/WixToolset.Core.WindowsInstaller/Bind/SequenceActionsCommand.cs:666]: I've removed it.

File: src/WixToolset.Core.WindowsInstaller/Bind/SequenceActionsCommand.cs [hunk 9]
```diff
@@ -562,64 +565,105 @@ private IEnumerable<string> GetRequiredActionIds()
...
+            Dictionary<string, WixActionSymbol> requiredActionSymbols,
+            Dictionary<WixActionSymbol, WixActionSymbol> firstReference)
         {
-            var after = false;
+            WixActionSymbol initialParentActionSymbol = null;
+            var previousActionSymbol = actionSymbol;
+            var initialActionSymbol = actionSymbol;
+            var initialAfter = this.GetAfter(actionSymbol);
 
-            if (actionSymbol.After != null)
```
[the remark this answers is not in the document]
Comment (cpuwzd) [reply src/WixToolset.Core.WindowsInstaller/Bind/SequenceActionsCommand.cs:573]: I have searched Core for "== null" and found 81 matching lines. None of them are ilines that I've changed.

File: src/WixToolset.Core.WindowsInstaller/Bind/SequenceActionsCommand.cs [hunk 10]
```diff
@@ -562,64 +565,105 @@ private IEnumerable<string> GetRequiredActionIds()
...
+                    firstReference[actionSymbol] = initialActionSymbol;
+                }
+                else if (firstReference[actionSymbol] == initialActionSymbol)
+                {
+                    throw new WixException(ErrorMessages.ActionCircularDependency(actionSymbol.SourceLineNumbers,
+                        actionSymbol.SequenceTable.ToString(), actionSymbol.Action, previousActionSymbol.Action));
+                }
+
+                if (null == actionSymbol.Before && null == actionSymbol.After)
+                {
```
[the remark this answers is not in the document]
Comment (cpuwzd) [reply src/WixToolset.Core.WindowsInstaller/Bind/SequenceActionsCommand.cs:605]: It's all been redone. The loop is very clean now.

File: src/WixToolset.Core.WindowsInstaller/Bind/SequenceActionsCommand.cs [hunk 11]
```diff
@@ -562,64 +565,105 @@ private IEnumerable<string> GetRequiredActionIds()
...
+                    }
 
-                    requiredActionSymbols.Add(parentActionSymbol.Id.Id, parentActionSymbol);
+                    break;
                 }
-                else
+
+                var parentActionSymbol = this.GetParentActionSymbol(actionSymbol, requiredActionSymbols, firstTime);
+
+                if (firstTime)
```
[the remark this answers is not in the document]
Comment (cpuwzd) [reply src/WixToolset.Core.WindowsInstaller/Bind/SequenceActionsCommand.cs:617]: It's all been redone.

What version of the compiler are you using? "??=" is not available before version 8. That's not what you environment sets up.

**Pull request closed without merging**